[core][nextjs] Implementation of separate component map for app router

MenKNas opened the pull request:
- [ ] This PR follows the [Contribution Guide](https://github.com/Sitecore/content-sdk/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation
The current implementation adds dual component map generation for Next.js App Router projects to prevent environment poisoning between server and client components.

Key Changes

- Dual Maps: Generates both component-map.ts (all components) and component-map.client.ts (client-safe only)
- Component Type Detection: Auto-detects 'use client' directives and explicit componentType exports ('server'|'client'|'universal')
- Router Auto-Detection: Automatically identifies App Router vs Pages Router from project structure
- Environment Separation: Server uses regular map, client uses client-safe map
- Backward Compatible: Pages Router projects continue to generate single map

## Testing Details
<!--- Please describe how you tested your changes. -->
<!--- When applicable, include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

MenKNas committed:
[core][nextjs] Implementation of seperate component map for client/server components in app router

MenKNas committed:
Merge branch 'dev' into feature/JSS-7735-generate-separate-component-maps-for-app-router

MenKNas committed:
Update CHANGELOG.md

MenKNas committed:
Update CHANGELOG.md

stasmaxymov commented on packages/core/src/tools/templating/components.ts:
```diff
@@ -72,3 +81,66 @@ export function getComponentList(paths: string[], exclude?: string[]): Component
+    const content = fs.readFileSync(filePath, 'utf-8');
+
+    // Check for 'use client' directive
+    if (content.includes("'use client'") || content.includes('"use client"')) {
```
I think simple "includes" check is not enough. Checking that first line is not complete either. I advise AST based checks, there are some example of similar stuff in the repo (e.g. import-map stuff)
It will also help with `componentType` export

stasmaxymov commented on packages/nextjs/src/tools/generate-map.ts:
```diff
@@ -18,23 +22,119 @@ export const generateMap: GenerateMapFunction = ({
   exclude,
   componentImports,
   mapTemplate = nextjsMapTemplate,
+  routerType,
```
I think explicit option to enable/disable client component map should be named like `clientComponentMap: boolean` instead of  `routerType`. Because you might want to use single component map in case you use only server or client components in project

stasmaxymov commented on packages/nextjs/src/tools/generate-map.ts:
```diff
@@ -18,23 +22,119 @@ export const generateMap: GenerateMapFunction = ({
+  components.forEach((component) => {
+    wildcardImports.push(`import * as ${component.moduleName} from '${component.importPath}';`);
+    // Add fallback componentType if not exported by the component
+    wildcardImports.push(
```
It seems these `componentType` assignments are mixed between imports which are going to look strange in component map. 
At minimum, we should separated them from imports section. 
But I even would argue that we don't need them in the current form. 

The only scenario where we need to explicitly set those is when component uses 'use client' directive, so we need only to modify such component entries and we can do that inline when add them to component map, e.g.:
`['Promo', {...Promo, componentType: 'client' }]`

faiqamin commented:
Looks good @MenKNas. Do we need to add some tests in here? Maybe to cover the map generation correctness or the classification rules?

stasmaxymov commented on packages/nextjs/src/tools/generate-map.ts:
```diff
@@ -18,23 +22,119 @@ export const generateMap: GenerateMapFunction = ({
   exclude,
   componentImports,
   mapTemplate = nextjsMapTemplate,
```
we have a way to override templating from cli.config, we probably need to extend it to work with client map too

stasmaxymov commented on packages/nextjs/src/tools/generate-map.ts:
```diff
@@ -18,23 +22,119 @@ export const generateMap: GenerateMapFunction = ({
   }
 };
 
+const nextjsMapTemplateWithTypes = (
```
Now we end-up with 3 copies of very similar templating functions, is there a way to combine them?

stasmaxymov reviewed (changes requested):
Overall looks good, please see some improvement comments

MenKNas committed:
Address PR comments

MenKNas committed:
Fix build issue and update test

MenKNas committed:
Fix linting issue

MenKNas commented:
Hello guys, PR update:

@faiqamin updated tests for `components.ts` to include new scenarios
@stasmaxymov addressed your comments and refactored some parts

MenKNas committed:
Remove duplicate ComponentType definition
